crates/sui-core/src/safe_client.rs

```diff
@@ -395,64 +395,89 @@ where
         if let (Some(e), Some(o)) = (expected_seq, observed_seq) {
             fp_ensure!(
                 e == o,
-                SuiError::ByzantineAuthoritySuspicion {
-                    authority: self.address,
-                }
+                SuiError::from("Expected checkpoint number doesn't match with returned")
             );
         }
         Ok(())
     }
 
-    pub async fn handle_checkpoint(
+    fn verify_contents_exist<T>(
         &self,
-        request: CheckpointRequest,
-    ) -> Result<CheckpointResponse, SuiError> {
-        let detail = request.detail;
-        let req_type = request.request_type.clone();
-
-        let resp = self.authority_client.handle_checkpoint(request).await?;
+        request_content: bool,
+        checkpoint: &Option<T>,
+        contents: &Option<CheckpointContents>,
+    ) -> SuiResult {
+        match (request_content, checkpoint, contents) {
+            // If content is requested, checkpoint is not None, but we are not getting any content,
+            // it's an error.
+            // If content is not requested, or checkpoint is None, yet we are still getting content,
+            // it's an error.
+            (true, Some(_), None) | (false, _, Some(_)) | (_, None, Some(_)) => Err(
+                SuiError::from("Checkpoint contents inconsistent with request"),
+            ),
+            _ => Ok(()),
+        }
+    }
 
+    fn verify_checkpoint_response(
+        &self,
+        request: &CheckpointRequest,
+        response: &CheckpointResponse,
+    ) -> SuiResult {
         // Verify signatures
-        resp.verify(&self.committee)?;
+        response.verify(&self.committee)?;
 
         // Verify response data was correct for request
-        match &req_type {
-            CheckpointRequestType::LatestCheckpointProposal => match &resp.info {
-                AuthorityCheckpointInfo::Proposal { previous, .. } => {
-                    self.verify_checkpoint_sequence(None, previous)?;
-                    Ok(resp)
+        match &request.request_type {
+            CheckpointRequestType::AuthenticatedCheckpoint(seq) => {
+                if let AuthorityCheckpointInfo::AuthenticatedCheckpoint(checkpoint) = &response.info
+                {
+                    // Checks that the sequence number is correct.
+                    self.verify_checkpoint_sequence(*seq, checkpoint)?;
+                    self.verify_contents_exist(request.detail, checkpoint, &response.detail)
```

Comment:
nit: I usually prefer to always add `?` and then return `Ok(())` at the end for uniformity.